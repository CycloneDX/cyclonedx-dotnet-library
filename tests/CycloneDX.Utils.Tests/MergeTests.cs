// This file is part of CycloneDX Library for .NET
//
// Licensed under the Apache License, Version 2.0 (the “License”);
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an “AS IS” BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) OWASP Foundation. All Rights Reserved.

using System;
using System.Collections.Generic;
using Xunit;
using Snapshooter;
using Snapshooter.Xunit;
using CycloneDX;
using CycloneDX.Models;
using CycloneDX.Models.Vulnerabilities;
using CycloneDX.Utils;

namespace CycloneDX.Utils.Tests
{
    public class MergeTests
    {
        [Fact]
        public void FlatMergeToolsTest()
        {
            #pragma warning disable 618
            var sbom1 = new Bom
            {
                Metadata = new Metadata
                {
                    Tools = new ToolChoices
                    {
                        Tools = new List<Tool>
                        {
                            new Tool
                            {
                                Name = "Tool1",
                                Version = "1"
                            }
                        }
                    }
                }
            };
            var sbom2 = new Bom
            {
                Metadata = new Metadata
                {
                    Tools = new ToolChoices
                    {
                        Tools = new List<Tool>
                        {
                            new Tool
                            {
                                Name = "Tool2",
                                Version = "1"
                            }
                        }
                    }
                }
            };
            #pragma warning restore 618

            var result = CycloneDXUtils.FlatMerge(sbom1, sbom2);

            Snapshot.Match(result);
        }

        [Fact]
        public void FlatMergeComponentsTest()
        {
            var sbom1 = new Bom
            {
                Components = new List<Component>
                {
                    new Component
                    {
                        Name = "Component1",
                        Version = "1"
                    }
                }
            };
            var sbom2 = new Bom
            {
                Components = new List<Component>
                {
                    new Component
                    {
                        Name = "Component2",
                        Version = "1"
                    }
                }
            };

            var result = CycloneDXUtils.FlatMerge(sbom1, sbom2);

            Snapshot.Match(result);
        }

        [Fact]
        public void FlatMergeDuplicatedComponentsTest()
        {
            var sboms = new List<Bom>();
            for (int i = 0; i < 3; i++)
            {
                var bom = new Bom
                {
                    Components = new List<Component>
                    {
                        new Component
                        {
                            Name = "Component1",
                            Version = "1"
                        }
                    }
                };
                sboms.Add(bom);
            }
            var result = CycloneDXUtils.FlatMerge(sboms);

            Assert.Single(result.Components);
        }


        [Fact]
        public void FlatMergeVulnerabilitiesTest()
        {
            var sbom1 = new Bom
            {
                Vulnerabilities = new List<Vulnerability>
                {
                    new Vulnerability
                    {
                        Id = "cve1",
                        Affects = new List<Affects>
                        {
                            new Affects
                            {
                                Ref = "ref1"
                            }
                        }
                    }
                }
            };
            var sbom2 = new Bom
            {
                Vulnerabilities = new List<Vulnerability>
                {
                    new Vulnerability
                    {
                        Id = "cve2",
                        Affects = new List<Affects>
                        {
                            new Affects
                            {
                                Ref = "ref2"
                            }
                        }
                    }
                }
            };

            var result = CycloneDXUtils.FlatMerge(sbom1, sbom2);

            Snapshot.Match(result);
        }

        [Fact]
        public void HierarchicalMergeComponentsTest()
        {
            var subject = new Component
            {
                Name = "Thing",
                Version = "1",
            };

            var sbom1 = new Bom
            {
                Metadata = new Metadata
                {
                    Component = new Component
                    {
                        Name = "System1",
                        Version = "1",
                        BomRef = "System1@1"
                    }
                },
                Components = new List<Component>
                {
                    new Component
                    {
                        Name = "Component1",
                        Version = "1",
                        BomRef = "Component1@1"
                    }
                },
                Dependencies = new List<Dependency>
                {
                    new Dependency
                    {
                        Ref = "System1@1",
                        Dependencies = new List<Dependency>
                        {
                            new Dependency
                            {
                                Ref = "Component1@1"
                            }
                        }
                    }
                },
                Compositions = new List<Composition>
                {
                    new Composition
                    {
                        Aggregate = Composition.AggregateType.Complete,
                        Assemblies = new List<string>
                        {
                            "System1@1"
                        },
                        Dependencies = new List<string>
                        {
                            "System1@1"
                        }
                    }
                }
            };
            var sbom2 = new Bom
            {
                Metadata = new Metadata
                {
                    Component = new Component
                    {
                        Name = "System2",
                        Version = "1",
                        BomRef = "System2@1"
                    }
                },
                Components = new List<Component>
                {
                    new Component
                    {
                        Name = "Component2",
                        Version = "1",
                        BomRef = "Component2@1"
                    }
                },
                Dependencies = new List<Dependency>
                {
                    new Dependency
                    {
                        Ref = "System2@1",
                        Dependencies = new List<Dependency>
                        {
                            new Dependency
                            {
                                Ref = "Component2@1"
                            }
                        }
                    }
                },
                Compositions = new List<Composition>
                {
                    new Composition
                    {
                        Aggregate = Composition.AggregateType.Complete,
                        Assemblies = new List<string>
                        {
                            "System2@1"
                        },
                        Dependencies = new List<string>
                        {
                            "System2@1"
                        }
                    }
                }
            };

            var result = CycloneDXUtils.HierarchicalMerge(new [] { sbom1, sbom2 }, subject);

            Snapshot.Match(result);
        }

        [Fact]
        public void HierarchicalMergeVulnerabilitiesTest()
        {
            var subject = new Component
            {
                Name = "Thing",
                Version = "1",
            };

            var sbom1 = new Bom
            {
                Metadata = new Metadata
                {
                    Component = new Component
                    {
                        Name = "System1",
                        Version = "1",
                        BomRef = "System1@1"
                    }
                },
                Vulnerabilities = new List<Vulnerability>
                {
                    new Vulnerability
                    {
                        Id = "cve1",
                        Affects = new List<Affects>
                        {
                            new Affects
                            {
                                Ref = "ref1"
                            }
                        }
                    }
                }
            };
            var sbom2 = new Bom
            {
                Metadata = new Metadata
                {
                    Component = new Component
                    {
                        Name = "System2",
                        Version = "1",
                        BomRef = "System2@1"
                    }
                },
                Vulnerabilities = new List<Vulnerability>
                {
                    new Vulnerability
                    {
                        Id = "cve2",
                        Affects = new List<Affects>
                        {
                            new Affects
                            {
                                Ref = "ref2"
                            }
                        }
                    }
                }
            };

            var result = CycloneDXUtils.HierarchicalMerge(new [] { sbom1, sbom2 }, subject);

            Snapshot.Match(result);
        }
    }
}
