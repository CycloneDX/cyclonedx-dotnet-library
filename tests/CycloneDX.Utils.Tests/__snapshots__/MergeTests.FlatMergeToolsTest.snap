{
  "BomFormat": "CycloneDX",
  "SpecVersion": "v1_6",
  "SpecVersionString": "1.6",
  "SerialNumber": null,
  "Version": null,
  "Metadata": {
    "Tools": {
      "Tools": [
        {
          "Vendor": null,
          "Name": "Tool1",
          "Version": "1",
          "Hashes": null
        },
        {
          "Vendor": null,
          "Name": "Tool2",
          "Version": "1",
          "Hashes": null
        }
      ]
    },
    "ProtobufTools": [
      {
        "Vendor": null,
        "Name": "Tool1",
        "Version": "1",
        "Hashes": null
      },
      {
        "Vendor": null,
        "Name": "Tool2",
        "Version": "1",
        "Hashes": null
      }
    ],
    "Authors": null,
    "Component": null,
    "Manufacture": null,
    "Supplier": null
  },
  "Components": null,
  "Compositions": null,
  "Definitions": null,
  "XmlSignature": null,
  "Signature": null
}
